syntax = "proto3";

package pb_common_types;

option go_package = "github.com/acompany-develop/QuickMPC/proto/common_types";

enum JobStatus {
  UNKNOWN = 0;
  ERROR = 1;
  RECEIVED = 2;
  PRE_JOB = 3;
  READ_DB = 4;
  COMPUTE = 5;
  COMPLETED = 6;
}

enum ComputationMethod {
  COMPUTATION_METHOD_UNSPECIFIED = 0;
  COMPUTATION_METHOD_MEAN = 1;
  COMPUTATION_METHOD_VARIANCE = 2;
  COMPUTATION_METHOD_SUM = 3;
  COMPUTATION_METHOD_CORREL = 4;
  COMPUTATION_METHOD_MESH_CODE = 5;
  COMPUTATION_METHOD_JOIN_TABLE = 6;
}

enum ShareValueTypeEnum {
    SHARE_VALUE_TYPE_UNSPECIFIED = 0;
    SHARE_VALUE_TYPE_FIXED_POINT = 1;
    SHARE_VALUE_TYPE_UTF_8_INTEGER_REPRESENTATION = 2;
}

message ProcedureProgress {
  uint64 id = 1;                // unique id in ProgressManager context
  string description = 2;       // describe specific process in job
  float progress = 3;           // describe progress by [0.0, 100.0]
  bool completed = 4;           // describe that procedure is completed or not
  optional string details = 5;  // describe progress in details
}

message JobProgress {
  string job_uuid = 1;
  JobStatus status = 2;
  repeated ProcedureProgress progresses = 3;
}

message Stacktrace {
  message Frame {
    string source_location = 1;
    uint64 source_line = 2;
    string function_name = 3;
  }
  repeated Frame frames = 1;
}

message JobErrorInfo {
  string what = 1;
  optional string additional_info = 2;
  optional Stacktrace stacktrace = 3;
}

message Schema {
    string name = 1;
    ShareValueTypeEnum type = 2;
}

message BigIntByte{
    bool sgn = 1;
    bytes abs_byte = 2;
}

